lib/spack/spack/environment/environment.py

```diff
@@ -1024,6 +1024,47 @@ def add(self, user_spec, list_name=user_speclist_name):
 
         return bool(not existing)
 
+    def change_existing_spec(self, change_spec, list_name=user_speclist_name, match_spec=None):
+        """
+        Provide a single spec and assume you want to find another spec with the
+        same name and override it with any conflicting details in change_spec.
+
+        This is only allowed for environments that do not use matrices (i.e.
+        they have a single, explicitly-enumerated list of specs called
+        'specs').
+        """
+        if not (change_spec.name or (match_spec and match_spec.name)):
+            raise ValueError(
+                "Must specify a spec name to identify a single spec"
+                " in the environment that will be changed"
+            )
+        match_spec = match_spec or Spec(change_spec.name)
+
+        list_to_change = self.spec_lists[list_name]
+        if list_to_change.is_matrix:
+            raise SpackEnvironmentError(
+                "Cannot directly change specs in matrices:"
+                " specify a named list that is not a matrix"
+            )
+
+        matches = list(x for x in list_to_change if x.satisfies(match_spec))
+        if len(matches) == 0:
+            raise ValueError(
+                "There are no specs named {0} in {1}".format(match_spec.name, list_name)
+            )
+        elif len(matches) > 1:
+            raise ValueError("{0} matches multiple specs".format(str(match_spec)))
```

Comment:
is it not safe to allow modifying multiple specs?

Reply:
A `-a` option has now been added.